app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataDao.java

```diff
@@ -121,19 +121,19 @@ public HealthDataRecordBuilder getRecordBuilder() {
         List<DynamoHealthDataRecord> recordList = mapper.query(DynamoHealthDataRecord.class, expression);
 
         List<HealthDataRecord> finalRecords = new ArrayList<>();
-        for (Object oneResult : recordList) {
-            if (!DynamoHealthDataRecord.class.isInstance(oneResult)) {
-                // This should never happen, but just in case.
-                throw new BridgeServiceException(String.format(
-                        "DynamoDB returned objects of type %s instead of %s",
-                        oneResult.getClass().getName(), DynamoHealthDataRecord.class.getName()));
+        int numDupes = 0;
+        for (DynamoHealthDataRecord oneResult : recordList) {
+            if (oneResult.getSchemaId().equals(schemaId)) {
+                finalRecords.add(oneResult);
+                numDupes++;
+
+                if (numDupes >= BridgeConstants.DUPE_RECORDS_MAX_COUNT) {
+                    // Limit the number of dupes we get to prevent browning out DDB.
+                    break;
+                }
             }
-
-            finalRecords.add((HealthDataRecord) oneResult);
         }
 
-        return finalRecords.stream()
-                .filter(record -> record.getSchemaId().equals(schemaId))
-                .collect(Collectors.toList());
```

Comment:
I think stream.limit(BridgeConstants.DUPE_RECORDS_MAX_COUNT) should also work here. The stream is supposed to be smart enough to stop after the limit number without processing the rest, and this is all happening in memory.